src/10-ipfs-0-4-0-released/index.md

```diff
@@ -0,0 +1,121 @@
+
+### On Breaking Changes
+
+This is a **breaking change** release, with a major refactor and upgrade to the
+IPFS networking protocol. Because of the refactor, **all IPFS daemons earlier
+than 0.4.0 will not be able to communicate with the newest version**. It is
+strongly recommended that everyone running an IPFS node upgrades to the latest
+version as soon as possible, as these nodes will, after a certain time, no
+longer be able to communicate with the majority of the network until they are
+upgraded.
+
+Refactoring the protocol is not something to be done lightly. But at this early
+stage, this is necessary to ensure we have the right design for IPFS in place
+for the future. It's better to improve the protocol now during this alpha stage
+of the project than when there are a lot more people running nodes across a
+lot of different implementations.
+
+One of the important changes that's been made to the protocol means that *there
+should never be a breaking change like this again*. This is due to a change to
+allow nodes to announce the version of the protocol they are using when
+connecting to other IPFS nodes. The goal is to roll any future protocol changes
+into the implementations gradually, so that we can still support legacy
+protocols for a period of time, making it easier to deprecate old versions over
+time.
+
+## Why we're changing the protocol
+
+The most important change was to allow IPFS implementations to use pluggable
+stream multiplexers, such as [yamux](https://github.com/hashicorp/yamux),
+[spdystream](https://github.com/docker/spdystream), or
+[muxado](https://github.com/inconshreveable/muxado). Instead of locking IPFS
+permanently into a single multiplexer that won't work for every language or
+situation, this change allows the implementations to implement the multiplexers
+of their choosing, and then negotiate which to use during the initial connection
+handshake.
+
+This modularity with stream muxing makes it easier for certain languages to
+improve performance. For example, the Go programming language may have muxado
+and yamux implementations that are really good, but many languages lack good (or
+any) implementations of them. For example, Node.js works well with spdystream in
+testing, and it would be nice to take advantage of that. And then there are
+options like 'multiplex', which may not have the same performance, but are much
+easier to implement.
+
+So by supporting as many muxers as we can, we get to choose the best
+multiplexers for the job. It also makes it much easier to implement the IPFS
+protocols in a new language. And of course, if a better multiplexers standard
+comes along, it will be easier to upgrade IPFS to support it in the future.
+
+The same code that allows us to easily select a stream muxer is also being
+used to select which ipfs sub-protocol to use for any given stream between
+peers. Now, if we need to make a breaking protocol change to any of those (like
```

Comment:
We won't need to 'break' anything anymore because it is ok to support multiple versions at the same time :)